File: 978-valid-mountain-array/valid-mountain-array.py
class Solution(object):
    def validMountainArray(self, arr):
        """
        :type arr: List[int]
        :rtype: bool
        """
        size = len(arr)
        peak = 0

        if size < 3:
            return False

        for index in range(1, size):
            if arr[index - 1] >= arr[index]:
                break
            peak = index
        
        if peak == size - 1 or peak == 0:
            return False
        
        for idx in range(peak + 1, size):
            if arr[idx] >= arr[idx - 1]:
                return False
        
        return True

        

File: 978-valid-mountain-array/valid-mountain-array.js
/**
 * @param {number[]} arr
 * @return {boolean}
 */
var validMountainArray = function(arr) {
    let n = arr.length;
    if (n < 3) return false;

    let i = 0;

    // 1️climb up
    while (i + 1 < n && arr[i] < arr[i + 1]) {
        i++;
    }

    // peak cannot be first or last
    if (i === 0 || i === n - 1) return false;

    // go down
    while (i + 1 < n && arr[i] > arr[i + 1]) {
        i++;
    }

    // must end exactly at last index
    return i === n - 1;
};